File: packages/core/content/source-inbox/contract.ts
import { z } from "zod";
import { parseWithFallback } from "@multica/core/api/schema";

// The material inbox's wire contract (specs/028, SOP §4).
//
// kind and status are open-ended strings on reads, deliberately. An older
// installed client receiving a value a newer server introduced renders the raw
// value rather than replacing the whole row with a fallback - the item and its
// annotation are still worth showing.

export const SOURCE_KINDS = ["pasted_text", "url"] as const;
export type SourceKind = (typeof SOURCE_KINDS)[number];

export const SOURCE_STATUSES = ["inbox", "organized", "archived"] as const;
export type SourceStatus = (typeof SOURCE_STATUSES)[number];

const sourceSchema = z
  .object({
    source_id: z.string(),
    workspace_id: z.string(),
    kind: z.string(),
    url: z.string(),
    captured_at: z.string(),
    recorded_by: z.string(),
    historical_import: z.boolean(),
    title: z.string(),
    tags: z.array(z.string()).nullable(),
    annotation: z.string(),
    personal_judgement: z.string(),
    status: z.string(),
    updated_at: z.string(),
  })
  .loose()
  .transform((source) => ({
    sourceId: source.source_id,
    workspaceId: source.workspace_id,
    kind: source.kind,
    url: source.url,
    capturedAt: source.captured_at,
    recordedBy: source.recorded_by,
    historicalImport: source.historical_import,
    title: source.title,
    // A null collection is normalised to an empty one so every reader can
    // map over it without knowing which of the two it received.
    tags: source.tags ?? [],
    annotation: source.annotation,
    personalJudgement: source.personal_judgement,
    status: source.status,
    updatedAt: source.updated_at,
  }));

export type Source = z.infer<typeof sourceSchema>;

const snapshotSchema = z
  .object({
    snapshot_id: z.string(),
    workspace_id: z.string(),
    source_id: z.string(),
    content: z.string(),
    content_hash: z.string(),
    captured_at: z.string(),
  })
  .loose()
  .transform((snapshot) => ({
    snapshotId: snapshot.snapshot_id,
    workspaceId: snapshot.workspace_id,
    sourceId: snapshot.source_id,
    content: snapshot.content,
    contentHash: snapshot.content_hash,
    capturedAt: snapshot.captured_at,
  }));

export type SourceSnapshot = z.infer<typeof snapshotSchema>;

const revisionSchema = z
  .object({
    revision_id: z.string(),
    workspace_id: z.string(),
    source_id: z.string(),
    changed_fields: z.array(z.string()).nullable(),
    actor_id: z.string(),
    created_at: z.string(),
  })
  .loose()
  .transform((revision) => ({
    revisionId: revision.revision_id,
    workspaceId: revision.workspace_id,
    sourceId: revision.source_id,
    changedFields: revision.changed_fields ?? [],
    actorId: revision.actor_id,
    createdAt: revision.created_at,
  }));

export type SourceRevision = z.infer<typeof revisionSchema>;

const bulkResultSchema = z
  .object({
    source_id: z.string(),
    ok: z.boolean(),
    reason: z.string().optional(),
  })
  .loose()
  .transform((result) => ({
    sourceId: result.source_id,
    ok: result.ok === true,
    reason: result.reason ?? "",
  }));

export type SourceBulkResult = z.infer<typeof bulkResultSchema>;

const EMPTY_SOURCE: Source = {
  sourceId: "",
  workspaceId: "",
  kind: "pasted_text",
  url: "",
  capturedAt: "",
  recordedBy: "",
  historicalImport: false,
  title: "",
  tags: [],
  annotation: "",
  personalJudgement: "",
  status: "inbox",
  updatedAt: "",
};

const sourceListSchema = z
  .object({ sources: z.array(sourceSchema) })
  .loose()
  .transform((list) => list.sources);

const revisionListSchema = z
  .object({ revisions: z.array(revisionSchema) })
  .loose()
  .transform((list) => list.revisions);

const duplicatesSchema = z
  .object({ source_ids: z.array(z.string()).nullable() })
  .loose()
  .transform((body) => body.source_ids ?? []);

const bulkSchema = z
  .object({ results: z.array(bulkResultSchema) })
  .loose()
  .transform((body) => body.results);

const detailSchema = z
  .object({
    source: sourceSchema,
    snapshot: snapshotSchema.nullable().optional(),
  })
  .loose()
  .transform((body) => ({ source: body.source, snapshot: body.snapshot ?? null }));

/**
 * What creating answers with.
 *
 * `duplicates` is a hint and nothing else: the item was created, and neither it
 * nor the ones it names were changed (R-011). A caller that treats a non-empty
 * list as a failure would be inventing a rule the SOP does not have.
 */
const createdSchema = z
  .object({
    source: sourceSchema,
    snapshot: snapshotSchema.nullable().optional(),
    duplicates: z.array(z.string()).nullable(),
  })
  .loose()
  .transform((body) => ({
    source: body.source,
    snapshot: body.snapshot ?? null,
    duplicates: body.duplicates ?? [],
  }));

export type SourceDetail = { source: Source; snapshot: SourceSnapshot | null };
export type CreatedSource = SourceDetail & { duplicates: string[] };

export function parseSources(data: unknown): Source[] {
  return parseWithFallback(data, sourceListSchema, [] as Source[], {
    endpoint: "GET /api/content-sources",
  });
}

export function parseSourceDetail(data: unknown): SourceDetail {
  return parseWithFallback(data, detailSchema, { source: EMPTY_SOURCE, snapshot: null }, {
    endpoint: "GET /api/content-sources/{id}",
  });
}

export function parseCreatedSource(data: unknown): CreatedSource {
  return parseWithFallback(
    data,
    createdSchema,
    { source: EMPTY_SOURCE, snapshot: null, duplicates: [] },
    { endpoint: "POST /api/content-sources" },
  );
}

export function parseSourceRevisions(data: unknown): SourceRevision[] {
  return parseWithFallback(data, revisionListSchema, [] as SourceRevision[], {
    endpoint: "GET /api/content-sources/{id}/revisions",
  });
}

export function parseDuplicates(data: unknown): string[] {
  return parseWithFallback(data, duplicatesSchema, [] as string[], {
    endpoint: "GET /api/content-sources/duplicates",
  });
}

export function parseBulkResults(data: unknown): SourceBulkResult[] {
  return parseWithFallback(data, bulkSchema, [] as SourceBulkResult[], {
    endpoint: "POST /api/content-sources/bulk",
  });
}

/** Whether a status is one this build knows how to label. */
export function isKnownStatus(status: string): status is SourceStatus {
  return (SOURCE_STATUSES as readonly string[]).includes(status);
}

export function isKnownKind(kind: string): kind is SourceKind {
  return (SOURCE_KINDS as readonly string[]).includes(kind);
}

File: server/internal/content/source-inbox/contract.go
// Package sourceinbox owns the manual half of SOP §4: taking something a person
// pasted or linked, keeping it with the facts about how it was collected, and
// letting them organise it later.
//
// Four things this package deliberately cannot do, each with a test:
//
//   - It makes no outbound request of any kind. A url source stores the link
//     and whatever the person typed about it; nothing here fetches the page
//     (spec FR-002).
//   - It calls no model. §4 step 4's "建议摘要、实体、观点、问题与关联资料" is
//     the parser's column, and there is no parser here.
//   - It never merges or deletes a duplicate. §4 says "重复素材先提示合并关联"
//     and R-011 says "内容相同不删除独立的收藏上下文与批注" - so this offers the
//     hint and stops there.
//   - It has no parse-status column. SOP §7.1 lists pending → processing →
//     ready / partial / failed; nothing here would advance it, and a column
//     frozen at one value is a decoration that reads as a claim.
//
// Contract: specs/028-source-inbox-manual/contracts/source-inbox.md
package sourceinbox

import (
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"net/url"
	"strings"
	"time"
	"unicode/utf8"
)

var (
	// ErrInvalid is unusable input: a value outside a controlled set, a
	// required field left out, an immutable column someone tried to change.
	// Answered as 400 with a diagnostic error object naming the field.
	ErrInvalid = errors.New("invalid source inbox input")
	// ErrNotFound is "no such source here". Answered identically to a refusal,
	// so a caller cannot learn from it whether the thing exists in another
	// workspace.
	ErrNotFound = errors.New("source not found")
	// ErrStorage is anything the database said. Never carries its text.
	ErrStorage = errors.New("source inbox storage unavailable")
)

// FieldError names the field that was wrong, so a 400 can say which one.
// "参数错误" tells the person filling the form nothing about what to fix.
type FieldError struct {
	Field  string
	Reason string
}

func (e FieldError) Error() string { return e.Reason + ": " + e.Field }
func (e FieldError) Unwrap() error { return ErrInvalid }

func invalidField(field string) error {
	return FieldError{Field: field, Reason: "invalid or missing"}
}

// Kind is what was collected. §4 step 1 lists pasting a link, a quick note,
// uploading a file and picking exported material; this card's inputs are the
// first two, and a quick note is a pasted text.
//
// Files and exports (W-03) are out of scope, so this set has two values and
// adding a third is a migration - the CHECK in 516 moves with it.
type Kind string

const (
	KindPastedText Kind = "pasted_text"
	KindURL        Kind = "url"
)

var Kinds = []Kind{KindPastedText, KindURL}

// Status is where an item is in being organised. It is NOT SOP §7.1's parse
// status: that one describes what a parser did to the material, this one
// describes what a person decided about it. §7.1 itself says "解析状态与知识是
// 否确认分开".
type Status string

const (
	StatusInbox     Status = "inbox"
	StatusOrganized Status = "organized"
	StatusArchived  Status = "archived"
)

var Statuses = []Status{StatusInbox, StatusOrganized, StatusArchived}

// MaxContentRunes bounds a pasted body. Over it the save is refused, never
// truncated: the hash is computed over what is stored, so truncating would
// silently give a hash to a body nobody meant to save - and that hash is the
// only thing duplicate detection has to go on.
const MaxContentRunes = 200000

const MaxFieldRunes = 4000

// Source is one act of collecting. The first five fields describe the event
// and never change; the rest are what organising touches.
type Source struct {
	SourceID          string    `json:"source_id"`
	WorkspaceID       string    `json:"workspace_id"`
	Kind              Kind      `json:"kind"`
	URL               string    `json:"url"`
	CapturedAt        time.Time `json:"captured_at"`
	RecordedBy        string    `json:"recorded_by"`
	HistoricalImport  bool      `json:"historical_import"`
	Title             string    `json:"title"`
	Tags              []string  `json:"tags"`
	Annotation        string    `json:"annotation"`
	PersonalJudgement string    `json:"personal_judgement"`
	Status            Status    `json:"status"`
	UpdatedAt         time.Time `json:"updated_at"`
}

// Snapshot is the material itself plus its hash. Append-only.
type Snapshot struct {
	SnapshotID  string    `json:"snapshot_id"`
	WorkspaceID string    `json:"workspace_id"`
	SourceID    string    `json:"source_id"`
	Content     string    `json:"content"`
	ContentHash string    `json:"content_hash"`
	CapturedAt  time.Time `json:"captured_at"`
}

// Revision is one act of organising. It names the fields that changed rather
// than copying the row.
type Revision struct {
	RevisionID    string    `json:"revision_id"`
	WorkspaceID   string    `json:"workspace_id"`
	SourceID      string    `json:"source_id"`
	ChangedFields []string  `json:"changed_fields"`
	ActorID       string    `json:"actor_id"`
	CreatedAt     time.Time `json:"created_at"`
}

// NewSource is what a caller may set when collecting. Note what is absent:
// there is no status (everything starts in the inbox) and no way to set
// captured_at or recorded_by, which the store fills from the request.
type NewSource struct {
	Kind              Kind
	URL               string
	Content           string
	Title             string
	Tags              []string
	Annotation        string
	PersonalJudgement string
	HistoricalImport  bool
}

// Organize is the mutable half, as a patch. A nil field is "leave it".
type Organize struct {
	Title             *string
	Tags              *[]string
	Annotation        *string
	PersonalJudgement *string
	Status            *Status
}

// ContentHash is what duplicate detection compares. sha256 over the exact
// bytes stored, hex encoded.
//
// No normalisation - not trimming, not case folding, not collapsing
// whitespace. Two bodies that differ by a trailing newline are two different
// texts, and deciding they are "the same" is exactly the merge judgement §4
// leaves to a person.
func ContentHash(content string) string {
	sum := sha256.Sum256([]byte(content))
	return hex.EncodeToString(sum[:])
}

// oneOf accepts an exact match only: no trimming, no case folding. Tolerating
// "URL" beside "url" would give one value two spellings in storage, and every
// reader would then have to know both.
func oneOf[T ~string](value string, allowed []T) bool {
	for _, candidate := range allowed {
		if string(candidate) == value {
			return true
		}
	}
	return false
}

func ValidateKind(value string) error {
	if !oneOf(value, Kinds) {
		return invalidField("kind")
	}
	return nil
}

func ValidateStatus(value string) error {
	if !oneOf(value, Statuses) {
		return invalidField("status")
	}
	return nil
}

// ValidateNew checks one collection request as a whole, because the rules are
// about combinations: a url without a link and a pasted text with one are both
// wrong, and neither field can tell on its own.
func ValidateNew(input NewSource) error {
	if err := ValidateKind(string(input.Kind)); err != nil {
		return err
	}
	switch input.Kind {
	case KindURL:
		if err := validateURL(input.URL); err != nil {
			return err
		}
		// A url source has no body here: there is nothing to snapshot until
		// something fetches the page, and this card fetches nothing. Accepting
		// a body would store text the link does not vouch for.
		if input.Content != "" {
			return invalidField("content")
		}
	case KindPastedText:
		if strings.TrimSpace(input.Content) == "" {
			return invalidField("content")
		}
		if utf8.RuneCountInString(input.Content) > MaxContentRunes {
			return FieldError{Field: "content", Reason: "too long"}
		}
		if input.URL != "" {
			return invalidField("url")
		}
	}
	return validateFields(input.Title, input.Annotation, input.PersonalJudgement, input.Tags)
}

func validateURL(value string) error {
	if value == "" {
		return invalidField("url")
	}
	parsed, err := url.Parse(value)
	if err != nil || parsed.Host == "" {
		return invalidField("url")
	}
	// http and https only. A javascript: or file: link is not a source a
	// person can go back and read, and storing one invites a page to follow it.
	if parsed.Scheme != "http" && parsed.Scheme != "https" {
		return invalidField("url")
	}
	return nil
}

func validateFields(title, annotation, judgement string, tags []string) error {
	for field, value := range map[string]string{
		"title": title, "annotation": annotation, "personal_judgement": judgement,
	} {
		if utf8.RuneCountInString(value) > MaxFieldRunes {
			return FieldError{Field: field, Reason: "too long"}
		}
	}
	for _, tag := range tags {
		if strings.TrimSpace(tag) == "" || utf8.RuneCountInString(tag) > 100 {
			return invalidField("tags")
		}
	}
	return nil
}

// ValidateOrganize checks a patch. Only the mutable half can appear here; the
// handler rejects an immutable column before this is reached, and the store's
// UPDATE never names one.
func ValidateOrganize(patch Organize) error {
	if patch.Status != nil {
		if err := ValidateStatus(string(*patch.Status)); err != nil {
			return err
		}
	}
	title, annotation, judgement := "", "", ""
	if patch.Title != nil {
		title = *patch.Title
	}
	if patch.Annotation != nil {
		annotation = *patch.Annotation
	}
	if patch.PersonalJudgement != nil {
		judgement = *patch.PersonalJudgement
	}
	var tags []string
	if patch.Tags != nil {
		tags = *patch.Tags
	}
	return validateFields(title, annotation, judgement, tags)
}

// ChangedFields names what a patch touches, in a fixed order so a test can
// compare without sorting.
func ChangedFields(patch Organize) []string {
	fields := []string{}
	if patch.Title != nil {
		fields = append(fields, "title")
	}
	if patch.Tags != nil {
		fields = append(fields, "tags")
	}
	if patch.Annotation != nil {
		fields = append(fields, "annotation")
	}
	if patch.PersonalJudgement != nil {
		fields = append(fields, "personal_judgement")
	}
	if patch.Status != nil {
		fields = append(fields, "status")
	}
	return fields
}
